### PR TITLE
Deep CORAL + MnistUsps for Deep bench

### PR DESCRIPTION
**Pull request by YanisLalou**

### CODE REVIEW

File: solvers/DeepCORAL.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        x = nn.MaxPool2d(2)(x)
+        x = self.dropout1(x)
+        x = torch.flatten(x, 1)
+        x = self.mlp(x)
+        x = nn.ReLU()(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = nn.LogSoftmax(dim=1)(x)
+
+        return output
```
Comment (antoinecollas) [inline solvers/DeepCORAL.py:87]: you should avoid being dependent on the input_shape. You can try something like this:

```Python
import torch
import torch.nn as nn
import torch.nn.functional as F

class ShallowConvNet(nn.Module):
    def __init__(self, n_classes):
        super().__init__()

        # Convolutional layers and first dropout in a ModuleList
        self.convs = nn.ModuleList([
            nn.Conv2d(3, 32, 3, 1),  # First convolutional layer
            nn.ReLU(),
            nn.Conv2d(32, 64, 3, 1),  # Second convolutional layer
            nn.ReLU(),
            nn.Dropout(0.25)  # First dropout layer
        ])
        
        # Global Average Pooling (this layer is fixed and doesn't depend on input size)
        self.global_pool = nn.AdaptiveAvgPool2d((1, 1))
        
        # MLP layers with ReLU and Dropout
        self.mlp = nn.ModuleList([
            nn.Linear(64, 128),  # First MLP layer
            nn.ReLU(),
            nn.Dropout(0.5),  # Dropout after the first MLP layer
            nn.Linear(128, n_classes)  # Output layer directly in MLP
        ])

    def forward(self, x, y=None):
        # Forward pass through convolutional layers and first dropout
        for layer in self.convs:
            x = layer(x)
        
        # Apply global average pooling
        x = self.global_pool(x)
        
        # Flatten the tensor
        x = torch.flatten(x, 1)
        
        # Forward pass through MLP layers
        for layer in self.mlp:
            x = layer(x)
        
        logits = x
        loss = None
        
        if y is not None:
            loss = F.cross_entropy(logits, y)
        
        return logits, loss
```

File: solvers/DeepCORAL.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        self.criterions = {
+            'supervised': SupervisedScorer(),
+            'deep_embedded_validation': DeepEmbeddedValidation(),
+        }
+
+        #model = ResNet50WithMLP(n_classes=n_classes, input_shape=input_shape)
+        model = ShallowConvNet(n_classes=n_classes, input_shape=input_shape)
+        net = DeepCoral(
+            model,
+            optimizer=Adadelta,  # Adadelta is now defined
```
Comment (antoinecollas) [inline solvers/DeepCORAL.py:116]: why not AdamW? (it's not a big deal though)

File: solvers/DeepCORAL.py (hunk 1 above)
Comment (antoinecollas) [reply solvers/DeepCORAL.py:87]: and when data are grayscale (e.g. MNIST) you repeat 3 times the grayscale to mimic RGB

File: datasets/mnist_usps_pca.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+from benchopt import BaseDataset, safe_import_context
```
Comment (antoinecollas) [inline datasets/mnist_usps_pca.py:1]: is `mnist_usps_pca.py` the old `mnist_usps.py` ? for next time, you can use `git mv` so that a renaming does not appear as a new file

**Review by antoinecollas — COMMENTED** [review]
For now, my main comment is that we now have an `input_shape` that appears everywhere. Is it to define the neural network?
If so, I have given you an implementation below that is invariant to the input shape. If this works, you can consider removing `input_shape` everywhere.

File: solvers/DeepCORAL.py (hunk 1 above)
Comment (antoinecollas) [reply solvers/DeepCORAL.py:87]: I also added `(x, y)` in the forward so that you directly get the loss fct. Not sure it fits well the rest of the code though.

Comment (YanisLalou) [reply solvers/DeepCORAL.py:87]: Why do you think we need to compute the loss here knowing that its going to be computed again later on ?

File: solvers/DeepCORAL.py (hunk 2 above)
Comment (YanisLalou) [reply solvers/DeepCORAL.py:116]: Just took the same opt as in the PyTorch example, but for the final implementation, yes we should use a better opt

File: datasets/mnist_usps_pca.py (hunk 3 above)
Comment (YanisLalou) [reply datasets/mnist_usps_pca.py:1]: yes, nice to know. Thanks

File: solvers/DeepCORAL.py (hunk 2 above)
Comment (tgnassou) [reply solvers/DeepCORAL.py:116]: Adam was not working when I tried last time, but never tried AdamW

File: solvers/DeepCORAL.py (hunk 1 above)
Comment (tgnassou) [reply solvers/DeepCORAL.py:87]: Yes I think no need to have the loss in the model

Comment (antoinecollas) [reply solvers/DeepCORAL.py:87]: then your training loop is easy:
```python
optimizer = torch.optim.AdamW(model.parameters(), lr=3e-4)
for i in range(N_EPOCHS):
    optimizer.zero_grad()
    logits, loss = model(x, y)
    print(f"loss: {loss.item()}")
    loss.backward()
    optimizer.step()
```

Comment (antoinecollas) [reply solvers/DeepCORAL.py:87]: it's small remark, my main point was more about the `input_shape`

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+        # Extract label and domain from the image path
+        label = os.path.basename(os.path.dirname(image_path))
+        domain = os.path.basename(os.path.dirname(os.path.dirname(image_path)))
+
+        image = Image.open(image_path).convert("RGB")
+        if self.transform:
+            image = self.transform(image)
+        
+        return image, label
```
Comment (antoinecollas) [inline utils.py:105]: it's better to use pathlib than os.path

File: objective.py [hunk 5]
```diff
@@ -105,26 +107,29 @@ def evaluate_result(self, cv_results, dict_estimators):
...
-            y_pred_test_source = estimator.predict(X_test_source)
-            y_pred_test_source_proba = estimator.predict_proba(X_test_source)
+            y_pred_test_source = estimator.predict(X_test_source, sample_domain = sample_domain_test_source)
+            y_pred_test_source_proba = estimator.predict_proba(X_test_source, sample_domain = sample_domain_test_source)
 
             # Target accuracy
-            y_pred_test_target = estimator.predict(X_test_target)
-            y_pred_test_target_proba = estimator.predict_proba(X_test_target)
+            y_pred_test_target = estimator.predict(X_test_target, sample_domain = sample_domain_test_target)
+            y_pred_test_target_proba = estimator.predict_proba(X_test_target, sample_domain = sample_domain_test_target)
```
Comment (antoinecollas) [inline objective.py:132]: Why must we add all these `sample_domain` since we have not done it before?

Comment (YanisLalou) [reply objective.py:132]: Tbh, I think it was a mistake to not add all these `sample_domain` from the beginning. We didn't get any error raised before because in the implementation of shallow methods, the `sample_domain` is inferred as `target` if empty, while its not the case with deep methods.
All that to say, I think we should add them

File: datasets/mnist_usps.py [hunk 6]
```diff
@@ -23,32 +27,81 @@ class Dataset(BaseDataset):
...
-        'n_samples_source, n_samples_target': [(3000, 3000)],
+        #'n_samples_source, n_samples_target': [(3000, 3000)],
+        'n_samples_source, n_samples_target': [(10000, 10000)],
         'source_target': [('MNIST', 'USPS'),
                           ('USPS', 'MNIST')],
         'random_state': [27],
     }
 
-    path_preprocessed_data = "data/digit_preprocessed.pkl"
-    url_preprocessed_data = "https://figshare.com/ndownloader/files/46363525?private_link=f15a4a7c81084815114b"
```
[the remark this answers is not in the document]
Comment (YanisLalou) [reply datasets/mnist_usps.py:33]: Yes, we directly take the raw images instead of preprocessed vectors. We don't need to use feature reduction techniques beforehand since the network will handle that itself.

[the remark this answers is not in the document]
Comment (antoinecollas) [reply datasets/mnist_usps.py:33]: So there is one mnist_usps for shallow methods and one for deep methods?

[the remark this answers is not in the document]
Comment (antoinecollas) [reply datasets/mnist_usps.py:33]: ok I had forgotten, there are mnist_usps and mnist_usps_pca

**Review by antoinecollas — APPROVED** [review]

File: solvers/DeepCORAL.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+with safe_import_context() as import_ctx:
+    from benchmark_utils.base_solver import DASolver
+    from skada.deep import DeepCoral
+    import torch
+    from torch import nn
+    from torch.optim import AdamW
+    from skada.metrics import SupervisedScorer, DeepEmbeddedValidation
+
+
+class ShallowConvNet(nn.Module):
```
Comment (tgnassou) [inline solvers/DeepCORAL.py:15]: These networks are not specific to DeepCoral. Maybe we can create a utils/architectures.py

File: solvers/DeepCORAL.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        x = self.feature_layer(x)
+        for layer in self.mlp:
+            x = layer(x)
+
+        # Apply log softmax to get log probabilities
+        logits = nn.LogSoftmax(dim=1)(x)
+
+        return logits
+
+class ShallowMLP(nn.Module):
```
Comment (tgnassou) [inline solvers/DeepCORAL.py:62]: same here

File: solvers/DeepCORAL.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        elif dataset_name in ['bci']:
+            # Use double precision for BCI data
+            model = ShallowMLP(input_dim=253, n_classes=n_classes).double()
+        else:
+            raise ValueError(f"Unsupported dataset: {dataset_name}")
+        
+        net = DeepCoral(
+            model,
+            optimizer=AdamW,
+            reg=1,
```
Comment (tgnassou) [inline solvers/DeepCORAL.py:116]: For the next, we have to add this parameter to the gridsearch

File: solvers/DeepCORAL.py [hunk 10]
```diff
@@ -49,7 +50,6 @@ def get_estimator(self, n_classes, device, dataset_name, **kwargs):
         net = DeepCoral(
             model,
             optimizer=AdamW,
-            reg=1,
```
Comment (tgnassou) [inline solvers/DeepCORAL.py]: `reg` is the DA parameter (the docs is not correct in skada). In this commit, you just removed it, but you didn't add it in the grid search, right?

Comment (YanisLalou) [reply solvers/DeepCORAL.py]: Oh my bad, it thought it was an old optimiser reg term
